core/dbt/include/global_project/macros/materializations/helpers.sql

```diff
@@ -72,3 +72,96 @@
+   
+   {% if on_schema_change not in ['sync', 'append', 'fail', 'ignore'] %}
+     {{ return(default_value) }}
+
+   {% else %}
+     {{ return(on_schema_change) }}
+   
+   {% endif %}
+
+{% endmacro %}
+
+{% macro get_column_names(columns) %}
+  {% set result = [] %}
+  
+  {% for col in columns %}
+    {{ result.append(col.column) }}
+  {% endfor %}
+  
+  {{ return(result) }}
+{% endmacro %}
+
+{% macro diff_columns(source_columns, target_columns) %}
+  {% set result = [] %}
+  {% set source_names = get_column_names(source_columns) %}
+  {% set target_names = get_column_names(target_columns) %}
+   
+   {# check whether the name attribute exists in the target, but dont worry about data type differences #}
+   {%- for col in source_columns -%} 
+     {%- if col.name not in target_names -%}
+      {{ result.append(col) }}
+      {%- endif -%}
+   {%- endfor -%}
+  
+  {{ return(result) }}
+{% endmacro %}
+
+{% macro check_for_schema_changes(source_relation, target_relation) %}
+  
+  {% set schema_changed = False %}
+  {%- set source_columns = adapter.get_columns_in_relation(source_relation) -%}
+  {%- set target_columns = adapter.get_columns_in_relation(target_relation) -%}
+  {%- set source_not_in_target = diff_columns(source_columns, target_columns) -%}
+  {%- set target_not_in_source = diff_columns(target_columns, source_columns) -%}
+
+  {% if source_not_in_target != [] %}
+    {% set schema_changed = True %}
+  {% elif target_not_in_source != [] %}
+    {% set schema_changed = True %}
+  {% endif %}
+
+  {{ return(schema_changed) }}
+
+{% endmacro %}
+
+{% macro sync_columns(source_relation, target_relation, on_schema_change='append') %}
+  
+  {%- set source_columns = adapter.get_columns_in_relation(source_relation) -%}
+  {%- set target_columns = adapter.get_columns_in_relation(target_relation) -%}
+  {%- set add_to_target_arr = diff_columns(source_columns, target_columns) -%}
+  {%- set remove_from_target_arr = diff_columns(target_columns, source_columns) -%}
+
+  {%- if on_schema_change == 'append' -%}
+    {%- for col in add_to_target_arr -%}
+       {%- set build_sql = 'ALTER TABLE ' + target_relation.database + '.' + target_relation.schema+'.'+target_relation.name + ' ADD COLUMN ' + col.name + ' ' + col.dtype -%}
+       {%- do run_query(build_sql) -%}
+    {%- endfor -%}
+    
+  {% elif on_schema_change == 'sync' %}
+    {%- for col in add_to_target_arr -%}
+       {%- set build_sql = 'ALTER TABLE ' + target_relation.database + '.' + target_relation.schema+'.'+target_relation.name + ' ADD COLUMN ' + col.name + ' ' + col.dtype -%}
+       {%- do run_query(build_sql) -%}
+    {%- endfor -%}
+
+    {%- for col in remove_from_target_arr -%}
+      {%- set build_sql = 'ALTER TABLE ' + target_relation.database + '.' + target_relation.schema+'.'+target_relation.name + ' DROP COLUMN ' + col.name -%}
+      {%- do run_query(build_sql) -%}
+    {%- endfor -%}
```

Comment:
1. These need to be adapter-dispatched macros, since the specific DML syntax is going to look different on different adapters/databases.
2. On many adapters, it's possible to add/remove multiple columns in a single `ALTER TABLE` statement. I think that's preferable, for concision in general, and for adapters that don't use transactions in particular—if any of the column additions/deletions fail, the whole operation will roll back. The only tricky thing is getting the commas right. E.g. in PostgreSQL:
```sql
alter table dbt_jcohen.incremental
    add column fun int,
    add column funny varchar(10),
    drop column yelly,
    drop column hilarious;
```

So I think it would make sense to add a new macro to [`macros/adapters/common`](https://github.com/fishtown-analytics/dbt/blob/develop/core/dbt/include/global_project/macros/adapters/common.sql):
```sql
{% macro alter_relation_add_remove_columns(relation, add_columns = none, remove_columns = none) -%}
  {{ return(adapter.dispatch('alter_relation_add_column')(relation, add_columns, remove_columns)) }}
{% endmacro %}

{% macro default__alter_relation_add_remove_columns(relation, add_columns = none, remove_columns = none) -%}
  {% set sql -%}
      alter {{ relation.type }} {{ relation }}
          {% for column in add_columns %}
            add column {{ column.name }} {{ column.data_type }} {{ ',' if not loop.last }}
          {% endfor %}
          {{ ',' if remove_columns %}
          {% for column in remove_columns %}
            drop column {{ column.name }}
          {% endfor %}
  {%- endset %}
  {{ return(run_query(sql)) }}
{% endmacro %}
```
Then the logic here can just look like:
```sql
{%- if on_schema_change == 'append' -%}
   {%- do alter_relation_add_remove_columns(target_relation, add_to_target_arr) -%}
{% elif on_schema_change == 'sync' %}
   {%- do alter_relation_add_remove_columns(target_relation, add_to_target_arr, remove_from_target_arr) -%}
{% endif %}
```

Reply:
Right on - will make these updates - I was going to ask how to correct this for each database provider, so thanks for commenting on that proactively!